File: apps/client/src/lib/users/dtos.ts
import * as v from "valibot"

// Entity View DTOs ----------------------------------------

export type UserRoleDTO = "admin"

export interface UserDTO {
	id: string
	name: string
	email: string
	role: string
}

// Create User DTOs ----------------------------------------

export const createUserDTOSchema = v.object({
	name: v.pipe(
		v.string(),
		v.minLength(1, "El nombre es obligatorio"),
		v.maxLength(255, "El nombre no puede tener más de 255 caracteres"),
	),
	email: v.pipe(v.string(), v.email("El correo electrónico no es válido")),
	password: v.pipe(
		v.string(),
		v.check(
			(pw) => !validatePassword(pw),
			"Mínimo 8 caracteres, una mayúscula, una minúscula, un número y un carácter especial",
		),
	),
	role: v.picklist(["admin"] as const),
})

export type CreateUserDTOSchema = typeof createUserDTOSchema
export type CreateUserDTO = v.InferInput<typeof createUserDTOSchema>

// Update User DTOs ----------------------------------------

export const updateUserDTOSchema = v.object({
	name: v.optional(
		v.pipe(
			v.string(),
			v.minLength(1, "El nombre es obligatorio"),
			v.maxLength(255, "El nombre no puede tener más de 255 caracteres"),
		),
	),
	email: v.optional(v.pipe(v.string(), v.email("El correo electrónico no es válido"))),
	password: v.optional(
		v.pipe(
			v.string(),
			v.check(
				(pw) => !validatePassword(pw),
				"Mínimo 8 caracteres, una mayúscula, una minúscula, un número y un carácter especial",
			),
		),
	),
	role: v.optional(v.picklist(["admin"] as const)),
})

export type UpdateUserDTOSchema = typeof updateUserDTOSchema
export type UpdateUserDTO = v.InferInput<typeof updateUserDTOSchema>

// Password validation ----------------------------------------

export const validatePassword = (pw: string) => {
	const missing: string[] = []

	if (pw.length < 8) missing.push("al menos 8 caracteres")
	if (!/[a-z]/.test(pw)) missing.push("una minúscula")
	if (!/[A-Z]/.test(pw)) missing.push("una mayúscula")
	if (!/[0-9]/.test(pw)) missing.push("un número")
	if (!/[^a-zA-Z0-9]/.test(pw)) missing.push("un carácter especial")
	if (pw.length > 255) missing.push("máximo 255 caracteres")

	if (missing.length > 0) {
		return `La contraseña debe tener ${missing.join(", ")}`
	}

	return null
}

File: apps/server/src/auth/users/dtos.rs
use crate::auth::{User, UserId, UserRole};

use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use validator::{Validate, ValidationError};

#[derive(Debug, Default, Validate, Deserialize)]
pub struct GetUsersQuery {
	#[validate(length(
		max = 255,
		message = "El atributo 'search' no puede tener más de 255 caracteres"
	))]
	pub search: Option<String>,
	pub role: Option<UserRole>,
}

#[derive(Debug, Validate, Deserialize)]
pub struct CreateUserDto {
	#[validate(length(
		min = 1,
		max = 255,
		message = "El atributo 'name' no puede tener más de 255 caracteres"
	))]
	pub name: String,

	#[validate(email(message = "El atributo 'email' debe ser un correo electrónico válido"))]
	pub email: String,

	#[validate(custom(function = "validate_password"))]
	pub password: String,
	pub role: UserRole,
}

pub(crate) fn validate_password(password: &str) -> Result<(), ValidationError> {
	let mut missing = Vec::new();

	if password.len() < 8 {
		missing.push("al menos 8 caracteres");
	}
	if !password.chars().any(|c| c.is_ascii_lowercase()) {
		missing.push("una letra minúscula");
	}
	if !password.chars().any(|c| c.is_ascii_uppercase()) {
		missing.push("una letra mayúscula");
	}
	if !password.chars().any(|c| c.is_ascii_digit()) {
		missing.push("un número");
	}
	if !password.chars().any(|c| !c.is_ascii_alphanumeric()) {
		missing.push("un carácter especial");
	}

	if password.len() > 255 {
		missing.push("no más de 255 caracteres");
	}

	if missing.is_empty() {
		Ok(())
	} else {
		let mut err = ValidationError::new("password");
		err.message = Some(format!("La contraseña debe tener {}", missing.join(", ")).into());
		Err(err)
	}
}

#[derive(Debug, Clone, Serialize, FromRow)]
#[serde(rename_all = "camelCase")]
pub struct UserView {
	pub id: UserId,
	pub name: String,
	pub email: String,
	pub role: UserRole,
}

#[derive(Debug, Validate, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateUserDto {
	#[validate(length(
		min = 1,
		max = 255,
		message = "El atributo 'name' no puede tener más de 255 caracteres"
	))]
	pub name: Option<String>,

	#[validate(email(message = "El atributo 'email' debe ser un correo electrónico válido"))]
	pub email: Option<String>,

	pub role: Option<UserRole>,

	#[validate(custom(function = "validate_password"))]
	pub password: Option<String>,
}

impl From<User> for UserView {
	fn from(user: User) -> Self {
		Self {
			id: user.id,
			name: user.name,
			email: user.email,
			role: user.role,
		}
	}
}
